### PR TITLE
Allow setting number of worker threads and other HTTP/2 setting params

### LINKED ISSUE #145: Make HTTP2 window size configurable
`http2` uses a default connection window size of [1MB per connection and 256KB per stream](https://github.com/kazu-yamamoto/http2/blob/b054a706efa40f1c4b3ee077f2ac509c19a29147/Network/HTTP2/Client/Run.hs#L56), which can cause data to pile up on a receiver's heap if they can't keep up. We've observed this in the stress tests. The standard fix for this is to reduce the window size. `http2` makes this configurable, although a brief test where we tried to reduce the window size on the client led to bus errors in the server. We should make this configurable in `grapesy` as well, and see if we can avoid/fix the bus errors.

### PR DESCRIPTION
**Pull request by FinleyMcIlwaine**
We now support setting the `SETTINGS_INITIAL_WINDOW_SIZE` and `SETTINGS_MAX_CONCURRENT_STREAMS` settings via `http2`. We also support setting the initial connection window size, and the number of server worker threads.

The default values we set for these help us avoid the window size exhaustion deadlock that resulted in
[#168](https://github.com/well-typed/grapesy/issues/168).

Resolves #145

### CODE REVIEW

File: src/Network/GRPC/Server/Settings.hs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      --
+      -- __Note__: If a lower 'serverInitialConnectionWindowSize' is desired, the
+      -- number of workers should be increased to avoid a potential HTTP\/2
+      -- control flow deadlock.
+      serverNumberOfWorkers :: Word
+
+      -- | HTTP\/2 settings
+    , serverHTTP2Settings :: HTTP2Settings
+    }
+  deriving (Show)
```
Comment (edsko) [inline src/Network/GRPC/Server/Settings.hs:27]: Does this really have to be a separate record that we pass in separate from the `ServerConfig`? I would prefer if we could just add these two values there, seems simpler.

File: src/Network/GRPC/Common/HTTP2Settings.hs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+module Network.GRPC.Common.HTTP2Settings
+  ( HTTP2Settings(..)
+  , defaultHTTP2Settings
+  ) where
+
+import Data.Default
+import Data.Word
+
+-- | HTTP\/2 settings
+data HTTP2Settings = HTTP2Settings {
```
Comment (edsko) [inline src/Network/GRPC/Common/HTTP2Settings.hs:14]: I like that you bundled all of these parameters into a separate record, feels nice and clean.

File: src/Network/GRPC/Common/HTTP2Settings.hs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+      -- | Maximum number of concurrent active streams
+      --
+      -- <https://datatracker.ietf.org/doc/html/rfc7540#section-5.1.2>
+      http2MaxConcurrentStreams :: Word32
+
+      -- | Initial window size for streams
+      --
+      -- <https://datatracker.ietf.org/doc/html/rfc7540#section-6.9.2>
+    , http2InitialStreamWindowSize :: Word32
```
Comment (edsko) [inline src/Network/GRPC/Common/HTTP2Settings.hs:23]: I wonder if we should call this `http2StreamWindowSize`; I find the _initial_ part a bit confusing. While it is true that the spec calls it this, and `http2` echoes the terminology, in the specific way that `http2` _uses_ this value it's not so much the _initial_ window as the _maximum_ window. In principle of course we could start at `X`, then at some point it decreases to say `X/2`, and then we send a window update for some `Y > X` -- but `http2` will never do that: it will only ever send window updates up to `X` again.

File: src/Network/GRPC/Common/HTTP2Settings.hs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      -- If the consumed window space of all streams exceeds this value, the
+      -- sender will stop sending data. Therefore, if this value is less than
+      -- @'http2MaxConcurrentStreams' * 'http2InitialStreamWindowSize'@, there
+      -- is risk of a control flow deadlock, since the connection window space
+      -- may be used up by streams that we are not yet processing before we have
+      -- received all data on the streams that we /are/ processing. To reduce
+      -- this risk, increase 'serverNumberOfWorkers'. See
+      -- <https://github.com/kazu-yamamoto/network-control/pull/4> for more
+      -- information.
+    , http2InitialConnectionWindowSize :: Word32
```
Comment (edsko) [inline src/Network/GRPC/Common/HTTP2Settings.hs:42]: Same comment here regarding "initial".

I'm not quite sure why you are listing this one as specific to `http2` (perhaps due to an old comment of mine in the `grapesy` codebase, which I wrote prior to understanding all of this? :smile: ). The connection window size is not specific to `http2`-the-package, it's part of the spec; I couldn't find a specific section about the "connection window", but we could refer to https://datatracker.ietf.org/doc/html/rfc7540#section-6.9.1 for example. While it is true that `http2` sets this up at the start of the connection, I think that's something we don't to mention here (I think that text probably came from my old comment).

I really like the second paragraph, very clear.

File: src/Network/GRPC/Common/HTTP2Settings.hs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+-- [Section 6.5.2 of the HTTP\/2
+-- specification](https://datatracker.ietf.org/doc/html/rfc7540#section-6.5.2)
+-- recommends that the @SETTINGS_MAX_CONCURRENT_STREAMS@ parameter be no smaller
+-- than 100 "so as not to unnecessarily limit parallelism", so we default to
+-- 128.
+--
+-- The default initial stream window size (corresponding to the
+-- @SETTINGS_INITIAL_WINDOW_SIZE@ HTTP\/2 parameter) is 256KB. See
+-- 'Network.GRPC.Server.ServerSettings' for more information regarding this
+-- parameter.
```
Comment (edsko) [inline src/Network/GRPC/Common/HTTP2Settings.hs:75]: Thinking out loud: I wonder why we want a larger stream window size. The default in the spec is 64 kB (https://datatracker.ietf.org/doc/html/rfc7540#section-6.9.2). Perhaps that would be just fine for us? It would reduce our connection window from 32 MB to 8 MB, which seems worthwhile at little cost? 

Maybe put another way: I don't have a good intuition for how that connection window size affects performance, but we should probably have good reason to override the default from the spec?

File: src/Network/GRPC/Server/Settings.hs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+      -- on the currently active HTTP\/2 streams
+      --
+      -- This setting is specific to the
+      -- [http2](https://hackage.haskell.org/package/http2) package's
+      -- implementation of the HTTP\/2 specification.
+      --
+      -- __Note__: If a lower 'serverInitialConnectionWindowSize' is desired, the
+      -- number of workers should be increased to avoid a potential HTTP\/2
+      -- control flow deadlock.
+      serverNumberOfWorkers :: Word
```
Comment (edsko) [inline src/Network/GRPC/Server/Settings.hs:22]: Is this value only relevant to the server? The client runs the same frame receiver loop right..? Do the incoming frames get handled differently?

File: src/Network/GRPC/Common/HTTP2Settings.hs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+      -- will be broadcast via a @WINDOW_UDPATE@ frame at the beginning of a new
+      -- connection.
+      --
+      -- If the consumed window space of all streams exceeds this value, the
+      -- sender will stop sending data. Therefore, if this value is less than
+      -- @'http2MaxConcurrentStreams' * 'http2InitialStreamWindowSize'@, there
+      -- is risk of a control flow deadlock, since the connection window space
+      -- may be used up by streams that we are not yet processing before we have
+      -- received all data on the streams that we /are/ processing. To reduce
+      -- this risk, increase 'serverNumberOfWorkers'. See
```
Comment (edsko) [inline src/Network/GRPC/Common/HTTP2Settings.hs:39]: Do the haddocks for `serverNumberOfWorkers` get resolved correctly here? You probably need to use a fully qualified name?

File: src/Network/GRPC/Client/Connection.hs [hunk 8]
```diff
@@ -623,6 +633,6 @@ runTCPClient Address{addressHost, addressPort} =
 -- See docs of 'confBufferSize', but importantly: "this value is announced
 -- via SETTINGS_MAX_FRAME_SIZE to the peer."
 --
--- Value of 4kB is taken from the example code.
+-- Value of 4KB is taken from the example code.
```
Comment (edsko) [inline src/Network/GRPC/Client/Connection.hs:636]: Hah, fair enough. TIL :) Though https://en.wikipedia.org/wiki/Kilobyte suggests `KiB` instead :)

File: src/Network/GRPC/Common/HTTP2Settings.hs (hunk 5 above)
Comment (FinleyMcIlwaine) [reply src/Network/GRPC/Common/HTTP2Settings.hs:75]: Agreed, I'll use the default from the spec for now and experiment while profiling

File: src/Network/GRPC/Server/Settings.hs (hunk 6 above)
Comment (FinleyMcIlwaine) [reply src/Network/GRPC/Server/Settings.hs:22]: The concept of worker threads is specific to servers in `http2`, so this makes sense here I believe

File: src/Network/GRPC/Server/Settings.hs (hunk 1 above)
Comment (FinleyMcIlwaine) [reply src/Network/GRPC/Server/Settings.hs:27]: Yes, I will combine these!

File: src/Network/GRPC/Common/HTTP2Settings.hs (hunk 3 above)
Comment (FinleyMcIlwaine) [reply src/Network/GRPC/Common/HTTP2Settings.hs:23]: I will remove the `initial` prefixes on these

File: src/Network/GRPC/Spec/Status.hs [hunk 9]
```diff
@@ -41,7 +41,7 @@ data GrpcError =
     -- | Invalid argument
     --
     -- The client specified an invalid argument. Note that this differs from
-    -- 'GrpcFailedPrecondition': 'GrpcInvalidArgumen'` indicates arguments that
```
Comment (edsko) [inline src/Network/GRPC/Spec/Status.hs:44]: Hah, wow, thanks for catching these!

**Comment by edsko** [comment]
Thanks for the update. I've pushed two more commits, one that avoids hardcoding the default of 8 into grapesy by doing

```
          HTTP2.numberOfWorkers =
              fromMaybe
                (HTTP2.numberOfWorkers HTTP2.defaultServerConfig)
                (fromIntegral <$> serverOverrideNumberOfWorkers)
```

instead of having hte figure 8 there (and similarly in other places), and one commit to fix some code layout (long lines).

**Comment by edsko** [comment]
Although with https://github.com/kazu-yamamoto/http2/pull/130 that "number of workers" might not be relevant at all anymore? I'll verify.
